### DISCUSSION TITLE
https://vlaaad.github.io/clj-vs-cli

### DISCUSSION THREAD

**Discussion by giscus** [vlaaad/vlaaad.github.io#14] [Posts]
# https://vlaaad.github.io/clj-vs-cli

An unfair heavily biased apples-to-oranges comparison… unless?

https://vlaaad.github.io/clj-vs-cli

**Comment by navid-zamani** [comment]
WTF… ”REPL“ is literally just a new word for ”CLI“, created by kids who grew up with colorful clickables and thought the CLI was not invented yet.

Read, eval, print, in a loop, is literally what a CLI does!  I feel like I’m taking crazy pills here!!

And ”IDE“ is just a fancy shell. So a ”REPL IDE“ is just a normal shell again. Yes, before iLuddite UIs dominated, programming/scripting was what end-users did and were expected to do and expected from themselves due to being educated on a school. The experts were the ones doing hex math in their heads, putting oscilloscopes onto buses, soldering things and watching the blinkenlights.
Pride in and entitlement to stupidity wasn’t in fashion yet.

I wonder what term they will have for this in 30 years, when kids run wasmChromeOS in webChrome in Chrome on Google Fucksya v599.0 on their sensory catatonization implants.

**Reply by mathmul** [reply]
Yeah, the way I see it is that their names are so vague they practically mean the same thing. With a CLI, you write a command, press Enter, it runs, you get feedback printed out, and then you can run another command in the same interface. With a REPL, you write an expression, press Enter, it runs, you get feedback printed out, and then you can run another expression in the same interface.

Oh, so then a command is something different from an expression? Again, it’s just confusing naming. In ZSH, you can do `echo $((1 + 2))` just the same as `print(1+2)` in any REPL of choice.

So what is the actual difference? As far as I understand (and please, someone correct me if I'm wrong), the only difference is that a shell (CLI) talks to your **OS**, so you're controlling your **machine**, and each command runs in its own subprocess - while a REPL talks to your **language runtime**, so you’re controlling the **language interpreter**, and all commands run in the same process and share state.
